### ISSUE TITLE
Very slow '[data-cy="xyz"]:visible' selector

### ISSUE THREAD

**Issue by JesseZomer** [cypress-io/cypress#26453]
### Current behavior

We have an application where the DOM can get pretty big. We use the [data-cy] attribute for most selectors and that works pretty great and fast. 

Although sometimes we have the same selector two times on the page, but one hidden (one for mobile, one for desktop for example). We solve this by using `[data-cy=xyz]:visible` however you can see this slows the tests way down. Sometimes up to a few seconds per selector. Also the test runner seems to get stuck while trying to find the element.

### Desired behavior

Using `:visible` in your selector should still be fast and the test runner shouldn't get stuck.

### Test code to reproduce

https://github.com/JesseZomer/cypress-slow-visible-selector

I've added 3 tests. The first two are fast like normal, but the third one is a lot slower. Not as slow as in my application, because the page is pretty simple I assume. You can however definitely see the test runner timer getting stuck and see that the test takes considerably longer than the first two. 

### Cypress Version

12.9.0

### Node version

v16.14.2

### Operating System

Pop!_ OS 21.10

### Other

We believe the regression happened somewhere in v12, because our ci runtimes doubled. However I've tried my repo with v11.2 and I see the same results, so not really sure about that.

**Comment by jordanpowell88** [comment]
@JesseZomer Thanks for creating this issue and for creating a reproducible example. Your example clearly shows the performance difference between the two queries. I'm going to route this to our app team where they may investigate further.

**Comment by DVersavel** [comment]
This is still an issue in 14.0.1. Do you have any information about this?

**Comment by JesseZomer** [comment]
I've just tried it with 14.0.1. The third test is still slower than the first two tests. The runner also still seems to freeze for a tiny bit (the example is still pretty fast unfortunately so its hard to notice). I've tested this by pressing "F" while the tests are running, so the tests search bar toggles on/off.

I've updated the repo to the latest version.

**Comment by mirobo** [comment]
@JesseZomer is it still an issue with 14.5.4? 😅

**Comment by JesseZomer** [comment]
Yes. I've updated the example repo. The difference in the example is low, but you can clearly see the :visible selector test takes longer + get stuck for a little while.
